File: part2/2-event-driven-programming/part3/Hello.py
'''
Creating a new SimObject
    Step 1: Create a Python class
    Step 2: Implement the C++
    Step 3: Register the SimObject and C++ file
    Step 4: (Re-)build gem5
    Step 5: Create a config script
'''

from m5.params import *
from m5.SimObject import SimObject

class Hello(SimObject):
    type = 'Hello'
    cxx_header = "tutorial/hello.hh"

    time_to_wait = Param.Latency('2us', "Time before firing the event")
    number_of_fires = Param.Int(10, "Number of times to fire the event before goodbye")

File: part2/2-event-driven-programming/part3/hello.cc
#include "tutorial/hello.hh"
#include "debug/HelloDebug.hh"

Hello::Hello(HelloParams *params) : 
    SimObject(params), 
    event([this]{processEvent();}, 
    name()), 
    latency(params->time_to_wait), 
    timesLeft(params->number_of_fires)
{
    // cout << "Hello World! From a SimObject!" << endl;
    // DPRINTF(HelloDebug, "Created the hello object\n");
}

Hello* 
HelloParams::create() {
    return new Hello(this);
}


void Hello::processEvent() {
    timesLeft--;
    DPRINTF(HelloDebug, "Hello world! Processing the event! %d left\n", timesLeft);

    if (timesLeft <= 0) {
        DPRINTF(HelloDebug, "Done firing!\n");
    } else {
        schedule(event, curTick() + latency);
    }
}

void Hello::startup() {
    schedule(event, latency);
}